File: programmers/kakao/블록이동하기.py
# 프로그래머스 블록 이동하기
# 21.05.18

import sys
from collections import deque
import copy

rd = sys.stdin.readline


def getNext(pos, graph):
    next = []
    pos = list(pos)
    preRow1, preCol1, preRow2, preCol2 = pos[0][0], pos[0][1], pos[1][0], pos[1][1]

    direction = [(0, -1), (0, 1), (-1, 0), (1, 0)]

    for dx, dy in direction:
        row1, col1, row2, col2 = preRow1 + dy, preCol1 + dx, preRow2 + dy, preCol2 + dx

        if graph[row1][col1] == 0 and graph[row2][col2] == 0:
            next.append({(row1, col1), (row2, col2)})

    # 가로
    if preRow1 == preRow2:
        for i in [-1, 1]:
            if graph[preRow1 + i][preCol1] == 0 and graph[preRow2 + i][preCol2] == 0:
                next.append({(preRow1, preCol1), (preRow1 + i, preCol1)})
                next.append({(preRow2, preCol2), (preRow2 + i, preCol2)})

    # 세로
    if preCol1 == preCol2:
        for i in [-1, 1]:
            if graph[preRow1][preCol1 + i] == 0 and graph[preRow2][preCol2 + i] == 0:
                next.append({(preRow1, preCol1), (preRow1, preCol1 + i)})
                next.append({(preRow2, preCol2), (preRow2, preCol2 + i)})

    return next


def solution(board):
    n = len(board)
    now = {(1, 1), (1, 2)}
    graph = [[1] * (n + 2) for _ in range(n + 2)]

    for i in range(n):
        for j in range(n):
            graph[i + 1][j + 1] = board[i][j]

    q = deque()
    visited = []
    q.append((now, 0))
    visited.append(now)

    while q:
        pre, dist = q.popleft()
        if (n, n) in pre:
            return dist

        nexts = getNext(pre, graph)

        for nextPos in nexts:
            if nextPos not in visited:
                q.append((nextPos, dist + 1))
                visited.append(nextPos)

    return 0

File: programmers/kakao/블록이동하기.swift
// 프로그래머스 블록 이동하기
// 21.05.20

import Foundation

struct Queue<T> {
  private var left = [T]()
  private var right = [T]()
  init() {}

  public var isEmpty: Bool {
    return left.isEmpty && right.isEmpty
  }

  public var count: Int {
    return left.count + right.count
  }

  public mutating func push(_ element: T) {
    right.append(element)
  }

  public mutating func pop() -> T? {
    guard !isEmpty else { return nil }

    if left.isEmpty {
      left = right.reversed()
      right.removeAll()
    }

    return left.removeLast()
  }

  public func peek() -> T? {
    return left.isEmpty ? right.first : left.last
  }
}

struct Point: Hashable {
  var x: Int
  var y: Int

  init(_ y: Int, _ x: Int) {
    self.x = x
    self.y = y
  }

  static func == (lhs: Point, rhs: Point) -> Bool {
    return lhs.x == rhs.x && lhs.y == rhs.y
  }

  func hash(into hasher: inout Hasher) {
    hasher.combine(x)
    hasher.combine(y)
  }
}

func getNextPoints(_ pre: inout Set<Point>, _ graph: inout [[Int]]) -> [Set<Point>] {
  var result = [Set<Point>]()
  let direction = [(0, -1), (0, 1), (-1, 0), (1, 0)]
  let pre1 = pre.removeFirst()
  let pre2 = pre.removeFirst()

  for (dy, dx) in direction {
    let row1 = pre1.y + dy
    let col1 = pre1.x + dx
    let row2 = pre2.y + dy
    let col2 = pre2.x + dx

    if graph[row1][col1] == 0 && graph[row2][col2] == 0 {
      let temp: Set = [Point(row1, col1), Point(row2, col2)]
      result.append(temp)
    }
  }

  // 가로
  if pre1.y == pre2.y {
    for i in [-1, 1] {
      if graph[pre1.y + i][pre1.x] == 0 && graph[pre2.y + i][pre2.x] == 0 {
        let rotate1: Set = [pre1, Point(pre1.y + i, pre1.x)]
        let rotate2: Set = [pre2, Point(pre2.y + i, pre2.x)]
        result.append(rotate1)
        result.append(rotate2)
      }
    }
  }

  // 세로
  if pre1.x == pre2.x {
    for i in [-1, 1] {
      if graph[pre1.y][pre1.x + i] == 0 && graph[pre2.y][pre2.x + i] == 0 {
        let rotate1: Set = [pre1, Point(pre1.y, pre1.x + i)]
        let rotate2: Set = [pre2, Point(pre2.y, pre2.x + i)]
        result.append(rotate1)
        result.append(rotate2)
      }
    }
  }

  return result
}

func solution(_ board:[[Int]]) -> Int {
  let n = board.count
  var graph = [[Int]](repeating: [Int](repeating: 1, count: n + 2), count: n + 2)

  for i in 0..<n {
    for j in 0..<n {
      graph[i + 1][j + 1] = board[i][j]
    }
  }

  var start: Set = [Point(1,1), Point(1,2)]
  var q = Queue<(Set<Point>, Int)>()
  var visited = Set<Set<Point>>()

  visited.insert(start)
  q.push((start, 0))

  while !q.isEmpty {
    var (now, dist) = q.pop()!

    if now.contains(Point(n, n)) {
      return dist
    }

    let nextPoints = getNextPoints(&now, &graph)

    for next in nextPoints {
      if !visited.contains(next) {
        visited.insert(next)
        q.push((next, dist + 1))
      }
    }
  }

  return 0
}

print(solution([[0, 0, 0, 1, 1],[0, 0, 0, 1, 0],[0, 1, 0, 1, 1],[1, 1, 0, 0, 1],[0, 0, 0, 0, 0]]))
